pandas/_libs/lib.pyx

```diff
@@ -2888,7 +2936,7 @@ def map_infer_mask(
     """
     cdef:
         Py_ssize_t i, n
-        ndarray result
+        ndarray[numeric_object_t] result
```

Comment:
can we narrow down what types are actually needed here?  is there a danger of types that aren't part of numeric_object_t?

Reply:
The test suite only uses bool and int64. I'm going to check calls to make sure that's guaranteed. Assuming it is, does something like

```
ctypedef fused bool_int64_t:
    uint8_t
    int64_t
```

in dtypes.pxd work?

Reply:
All calls are either hard coded as bool, int64, or object. The only part that was a little tricky was the use in the various `_str_map` methods. Here, specifying the argument `dtype=np.dtype(dtype)` in `map_infer_mask` is guarded by  `if is_integer_dtype(dtype) or is_bool_dtype(dtype)`. All calls to the `_str_map` method are either a string dtype, int64, or bool.

Reply:
If we create the type here what would we name it? I agree its good to specialize instead of using `numeric_object_t` but it would be nice if the name could speak to the concept of what the type supports instead of being a literal declaration of its member types

Reply:
I agree with the sentiment but have no recommendations for a "natural" name. If we are not happy with the current name `uint8_int64_object_t`, then my only alternative is to move it from `dtypes.pxd` to `lib.pyx` just above the `map_infer_mask` function and name it `map_infer_mask_t` to highlight its local use.

Reply:
> current name `uint8_int64_object_t` [...] `map_infer_mask_t` to highlight its local use.

im fine with either of these options

Reply:
@WillAyd - friendly ping on `uint8_int64_object_t` vs `map_infer_mask_t` vs `something else`

Reply:
Let's go with the first option for now. Can always rename later